dask_ml/model_selection/_incremental.py

```diff
@@ -0,0 +1,277 @@
+    # This assumes that it fits into memory on a single worker
+    if isinstance(X_test, da.Array):
+        X_test = client.compute(X_test)
+    else:
+        y_test = yield client.scatter(y_test)
+    if isinstance(y_test, da.Array):
+        y_test = client.compute(y_test)
+    else:
+        y_test = yield client.scatter(y_test)
+
+    # Convert to batches of delayed objects of numpy arrays
+    X_train = X_train.to_delayed()
+    if hasattr(X_train, "squeeze"):
+        X_train = X_train.squeeze()
+    y_train = y_train.to_delayed()
+    if hasattr(y_train, "squeeze"):
+        y_train = y_train.squeeze()
+    X_train, y_train = dask.optimize(X_train.tolist(), y_train.tolist())
+
+    # Order by which we process training data futures
+    order = []
+    seen = {}
+    tokens = {}
+
+    def get_futures(time_step):
+        """ Policy to get training data futures
+
+        Currently we compute once, and then keep in memory.
+        Presumably in the future we'll want to let data drop and recompute.
+        This function handles that policy internally, and also controls random
+        access to training data.
+        """
+        # Shuffle blocks going forward to get uniform-but-random access
+        while time_step >= len(order):
+            L = list(range(len(X_train)))
+            rng.shuffle(L)
+            order.extend(L)
+
+        j = order[time_step]
+
+        if j in seen:
+            x_key, y_key = seen[j]
+            return Future(x_key), Future(y_key)
+        else:
+            # new future, need to tell scheduler about it
+            X_future = client.compute(X_train[j])
+            y_future = client.compute(y_train[j])
+            seen[j] = (X_future.key, y_future.key)
+
+            # Hack to keep the futures in the scheduler but not in memory
+            X_token = client.submit(len, X_future)
+            y_token = client.submit(len, y_future)
+            tokens[j] = (X_token, y_token)
+            return X_future, y_future
+
+    # Submit initial partial_fit and score computations on first batch of data
+    X_future, y_future = get_futures(0)
+    for ident, model in models.items():
+        model = client.submit(_partial_fit, model, X_future, y_future, fit_params)
+        score = client.submit(_score, model, X_test, y_test, scorer)
+        models[ident] = model
+        scores[ident] = score
+
+    seq = as_completed(scores.values(), with_results=True)
+    speculative = dict()  # models that we might or might not want to keep
+    history = []
+    number_to_complete = len(models)
+
+    # async for future, result in seq:
+    while not seq.is_empty():
+        future, meta = yield seq.__anext__()
+        if future.cancelled():
+            continue
+        ident = meta["ident"]
+
+        info[ident].append(meta)
+        history.append(meta)
+
+        # Evolve the model one more step
+        model = models[ident]
+        X_future, y_future = get_futures(meta["time_step"] + 1)
```

Comment:
I think we should replace `meta['time_step'] + 1` with `np.random.choice(X.blocks)` instead.

I think data access should be random, both across models and across time.

Reply:
Never mind about random access. `get_futures` manages that.